8345125: Aarch64: Add aarch64 backend for Float16 scalar operations

Bhavana-Kilambi opened the pull request:
This patch adds aarch64 backend for scalar FP16 operations namely - add, subtract, multiply, divide, fma, sqrt, min and max.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8345125](https://bugs.openjdk.org/browse/JDK-8345125): Aarch64: Add aarch64 backend for Float16 scalar operations (**Enhancement** - P4)


### Reviewers
 * [Andrew Haley](https://openjdk.org/census#aph) (@theRealAph - **Reviewer**)
 * [Hao Sun](https://openjdk.org/census#haosun) (@shqking - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/23748/head:pull/23748` \
`$ git checkout pull/23748`

Update a local copy of the PR: \
`$ git checkout pull/23748` \
`$ git pull https://git.openjdk.org/jdk.git pull/23748/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 23748`

View PR using the GUI difftool: \
`$ git pr show -t 23748`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/23748.diff">https://git.openjdk.org/jdk/pull/23748.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/23748#issuecomment-2678242999)
</details>

theRealAph commented on src/hotspot/cpu/aarch64/aarch64.ad:
```diff
@@ -17116,6 +17250,64 @@ instruct expandBitsL_memcon(iRegINoSp dst, memory8 mem, immL mask,
+  ins_pipe(pipe_slow);
+%}
+
+// This pattern would result in the following instructions (the first two are for ConvF2HF
+// and the last instruction is for ReinterpretS2HF) -
```
```suggestion
// Without this pattern, (ReinterpretS2HF (ConvF2HF src)) would result in the following instructions (the first two for ConvF2HF
// and the last instruction for ReinterpretS2HF) -
```
Reads a little better, I think?

Bhavana-Kilambi replied:
Addressed this in the new patch.

theRealAph commented on src/hotspot/cpu/aarch64/aarch64.ad:
```diff
@@ -6943,6 +6972,21 @@ instruct loadConD(vRegD dst, immD con) %{
+// Load Half Float Constant
+// ldr instruction has 32/64/128 bit variants but not a 16-bit variant. This
+// loads the 16-bit value from constant pool into a 32-bit register but only
+// the bottom half will be populated.
```
Surely what actually happens here is that it loads a 32-bit word from the constant pool. The bottom 16 bits of this word contain the half-precision constant.

Bhavana-Kilambi replied:
I agree. The wording didn't quite convey that. I will change it in my next PS. Thank you for looking into the patch!

Bhavana-Kilambi replied:
Addressed this in the new patch.

theRealAph commented on test/hotspot/gtest/aarch64/aarch64-asmtest.py:
```diff
@@ -14,10 +14,9 @@
-        0xff03, 0xfffe]
+        0x7e0, 0xfc0, 0x1f80, 0x3ff0, 0x7e00, 0x8000,
+        0x81ff, 0xc1ff, 0xc003, 0xc7ff, 0xdfff, 0xe03f,
+        0xe1ff, 0xf801, 0xfc00, 0xfc07, 0xff03, 0xfffe]
```
So here you've deleted the duplicated `0x7e00` (good) but also the not-duplicated `0xe10f`. Is `0xe10f` not valid?

theRealAph commented:
Overall, this looks like a great pice of work. I only have a few changes in comments and a question, then we're good to go.

shqking commented on src/hotspot/cpu/aarch64/assembler_aarch64.hpp:
```diff
@@ -2059,38 +2061,47 @@ void mvnw(Register Rd, Register Rm,
+  INSN(fnmuld, 0b000, 0b01, 0b100010, 0b1);
+
+  // Half-precision floating-point instructions
+  INSN(fabdh,  0b011, 0b11, 0b000101, 0b0);
```
I suppose `fadbh` and `fnmulh` are added to keep aligned with the float and double ones, i.e. `fabd(s|d)` and `fnmul(s|d)`.


I noticed that there are matching rules for `fabd(s|d)`, i.e. `absd(F|D)_reg`. I wonder if we need add the corresponding rule for fp16 here?

Bhavana-Kilambi replied:
Hi @shqking , thanks for your review comments. Yes I added `fabdh` and `fnmulh` to keep aligned with float and double types.
For adding support for FP16 `absd` we need `AbsHF` to be supported (along with SubHF) but `AbsHF` node is not implemented currently. `abs` operation is directly executed from the java code here - https://github.com/openjdk/jdk/blob/037e47112bdf2fa2324f7c58198f6d433f17d9fd/src/jdk.incubator.vector/share/classes/jdk/incubator/vector/Float16.java#L1464 and is not intrinsified or pattern matched like other FP16 operations. Same with `negate` operation for FP16 - https://github.com/openjdk/jdk/blob/037e47112bdf2fa2324f7c58198f6d433f17d9fd/src/jdk.incubator.vector/share/classes/jdk/incubator/vector/Float16.java#L1449
On the Valhalla repo, while these operation were being developed, I tried adding support for `AbsHF/NegHF` which emitted `fabs` and `fneg` instructions but the performance with the direct java code(bit manipulation operations) was much faster (sorry don't remember the exact number) so we decided to go with the java implementation instead.
I still added `fabd` here because `op21` is 0 only in `fabd` H variant and felt that it'd be better to handle it here as it belongs to this group of instructions. Please let me know your thoughts.

theRealAph replied:
According to the RM, fabd is in _Advanced SIMD scalar three same FP16_, but the rest are in _Floating-point data-processing (2 source)_.  The decoding scheme looks rather different.`fabd`, then, doesn't really fit here, but in a section with the rest of the three same FP16 instructions.
The encoding scheme for _Advanced SIMD scalar three same FP16_ is pretty simple, so I suggest you create a new group for them, and put `fabd` in there.

hsun-code replied:
@Bhavana-Kilambi Thanks for your explanation about the missing `AbsHF`. It's okay to me to have `fadbh` and `fnmulh` in this patch.

Overall LGTM except aph's comment above.

Bhavana-Kilambi replied:
Hi @theRealAph Thanks again for the review and apologies for the delay in responding.
I moved the three `fabd` instructions out of their current place and added them in two separate sections -  for the single and double precision (_Advanced SIMD scalar three same_) and another for the FP16 (_Advanced SIMD scalar three same FP16_). Please review the changes. Thank you!

Bhavana-Kilambi commented:
Hello @shqking @theRealAph , sincere apologies for the delay in addressing the review comments. I am planning on uploading a patch soon addressing all review comments. Thank you !

Bhavana-Kilambi commented:
Hello, I would not be able to respond to comments until the next couple months or so due to some urgent tasks at work. Until then, I'd move this PR to draft status so that it would not be closed due to lack of activity. Thank you for the review!

mhaessig commented:
Can you please uncomment the following tests using aarch64 float16 in this PR?

https://github.com/mhaessig/jdk/blob/cec48ed270d3bdf704c389a091b42a32c2ed6440/test/hotspot/jtreg/compiler/floatingpoint/TestSubNodeFloatDoubleNegation.java#L58-L60

Bhavana-Kilambi commented:
> Can you please uncomment the following tests using aarch64 float16 in this PR?
> 
> https://github.com/mhaessig/jdk/blob/cec48ed270d3bdf704c389a091b42a32c2ed6440/test/hotspot/jtreg/compiler/floatingpoint/TestSubNodeFloatDoubleNegation.java#L58-L60

Done. Thanks for notifying.

shqking commented:
Hi @Bhavana-Kilambi, I noticed there exists inconsistency between `test/hotspot/gtest/aarch64/asmtest.out.h` and `test/hotspot/gtest/aarch64/aarch64-asmtest.py` in the latest commit.  We should resolve that.

Bhavana-Kilambi commented:
> Hi @Bhavana-Kilambi, I noticed there exists inconsistency between `test/hotspot/gtest/aarch64/asmtest.out.h` and `test/hotspot/gtest/aarch64/aarch64-asmtest.py` in the latest commit. We should resolve that.

Thanks. I think I missed generating the `asmtest.out.h` file in the new commit. I'll update it.

Bhavana-Kilambi commented:
Hi @shqking I have regenerated the `test/hotspot/gtest/aarch64/asmtest.out.h` to keep it consistent with the instructions in test/`hotspot/gtest/aarch64/aarch64-asmtest.py`. Can you please review now? Thanks!

theRealAph reviewed (approved):
Looks good to me. Please summarize the tests that you've run before committing.

Bhavana-Kilambi commented:
> Looks good to me. Please summarize the tests that you've run before committing.

Thank you for the approval. 
All hotspot (hotspot_all), jdk (tiers 1-3) and langtools (tier1) pass on N1, V1 and V2 architectures.

shqking reviewed (approved):
Thanks for your update. Looks good to me.

theRealAph commented:
On 4/25/25 10:17, bkilambi wrote:
> Bhavana-Kilambi left a comment (openjdk/jdk#23748)
> 
>> Looks good to me. Please summarize the tests that you've run before committing.
> 
> Thank you for the approval.
> All hotspot (hotspot_all), jdk (tiers 1-3) and langtools (tier1) pass on N1, V1 and V2 architectures.

But do any of these tests use Float16 ?

-- 
Andrew Haley  (he/him)
Java Platform Lead Engineer
Red Hat UK Ltd. <https://www.redhat.com>
https://keybase.io/andrewhaley
EAC8 43EB D3EF DB98 CC77 2FAD A5CD 6035 332F A671

Bhavana-Kilambi commented:
> On 4/25/25 10:17, bkilambi wrote: Bhavana-Kilambi left a comment ([openjdk/jdk#23748](https://github.com/openjdk/jdk/pull/23748)) > Looks good to me. Please summarize the tests that you've run before committing. Thank you for the approval. All hotspot (hotspot_all), jdk (tiers 1-3) and langtools (tier1) pass on N1, V1 and V2 architectures.
> But do any of these tests use Float16 ?
> […](#)
> -- Andrew Haley (he/him) Java Platform Lead Engineer Red Hat UK Ltd. <https://www.redhat.com> https://keybase.io/andrewhaley EAC8 43EB D3EF DB98 CC77 2FAD A5CD 6035 332F A671

Yes they do.
The following FP16 tests are included in them and have also been tested separately (these are the ones which have been added/modified in the commit which added initial scalar FP16 support in JDK mainline [https://github.com/openjdk/valhalla/commit/f03fb4e4ee4d59ed692d0c26ddce260511f544e7] - 
```

test/hotspot/jtreg/compiler/c2/irTests/ConvF2HFIdealizationTests.java
test/hotspot/jtreg/compiler/c2/irTests/MulHFNodeIdealizationTests.java
test/hotspot/jtreg/compiler/c2/irTests/TestFloat16ScalarOperations.java
test/hotspot/jtreg/compiler/vectorization/TestFloat16VectorConvChain.java
test/jdk/jdk/incubator/vector/ScalarFloat16OperationsTest.java
```

Please let me know if you'd like me to test anything else. Thanks!

Bhavana-Kilambi commented:
Hi @theRealAph , is it ok if I integrate this or would you like me to do any other testing on the patch?

theRealAph commented:
> Hi @theRealAph , is it ok if I integrate this or would you like me to do any other testing on the patch?

As long as you've got test coverage for everything here, go ahead.

Bhavana-Kilambi commented:
Thanks a lot! Can I ask you to please sponsor this patch? 
/integrate

shqking commented:
/sponsor

TobiHartmann commented:
This caused a regression: [JDK-8355708](https://bugs.openjdk.org/browse/JDK-8355708)
